File: app/src/main/java/org/andstatus/app/actor/ActorContextMenuItem.java
/*
 * Copyright (C) 2016 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.andstatus.app.actor;

import android.net.Uri;
import android.view.Menu;

import org.andstatus.app.ActivityRequestCode;
import org.andstatus.app.MyAction;
import org.andstatus.app.account.AccountSelector;
import org.andstatus.app.account.MyAccount;
import org.andstatus.app.data.MatchedUri;
import org.andstatus.app.list.ContextMenuItem;
import org.andstatus.app.net.social.Actor;
import org.andstatus.app.note.NoteEditorData;
import org.andstatus.app.origin.Origin;
import org.andstatus.app.os.AsyncTaskLauncher;
import org.andstatus.app.os.MyAsyncTask;
import org.andstatus.app.service.CommandData;
import org.andstatus.app.service.CommandEnum;
import org.andstatus.app.service.MyServiceManager;
import org.andstatus.app.timeline.TimelineActivity;
import org.andstatus.app.timeline.meta.TimelineType;
import org.andstatus.app.util.MyLog;

public enum ActorContextMenuItem implements ContextMenuItem {
    GET_ACTOR(true) {
        @Override
        NoteEditorData executeAsync(Params params) {
            CommandData commandData = CommandData.newActorCommand(CommandEnum.GET_ACTOR,
                    params.menu.getViewItem().getActorId(),
                    params.menu.getViewItem().actor.getUsername());
            MyServiceManager.sendManualForegroundCommand(commandData);
            return super.executeAsync(params);
        }
    },
    PRIVATE_NOTE(true) {
        @Override
        NoteEditorData executeAsync(Params params) {
            return NoteEditorData.newEmpty(params.menu.getActingAccount())
                    .addToAudience(params.menu.getViewItem().getActorId());
        }

        @Override
        void executeOnUiThread(ActorContextMenu menu, NoteEditorData editorData) {
            if (editorData.activity.getNote().audience().hasNonPublic()) {
                menu.menuContainer.getNoteEditor().startEditingNote(editorData);
            }
        }
    },
    SHARE() {
        @Override
        void executeOnUiThread(ActorContextMenu menu, NoteEditorData editorData) {
            // TODO
        }
    },
    NOTES_BY_ACTOR(true) {
        @Override
        NoteEditorData executeAsync(Params params) {
            TimelineActivity.startForTimeline(
                    params.menu.getActivity().getMyContext(),
                    params.menu.getActivity(),
                    params.menu.getActivity().getMyContext().timelines()
                            .forUserAtHomeOrigin(TimelineType.SENT, params.menu.getViewItem().getActor()),
                    false);
            return super.executeAsync(params);
        }
    },
    FOLLOW() {
        @Override
        void executeOnUiThread(ActorContextMenu menu, NoteEditorData editorData) {
            sendActOnActorCommand(CommandEnum.FOLLOW, menu);
        }
    },
    STOP_FOLLOWING() {
        @Override
        void executeOnUiThread(ActorContextMenu menu, NoteEditorData editorData) {
            sendActOnActorCommand(CommandEnum.UNDO_FOLLOW, menu);
        }
    },
    ACT_AS_FIRST_OTHER_ACCOUNT() {
        @Override
        void executeOnUiThread(ActorContextMenu menu, NoteEditorData editorData) {
            menu.setSelectedActingAccount(menu.getMyContext().accounts()
                    .firstOtherSucceededForSameUser(menu.getViewItem().actor, menu.getActingAccount()));
            menu.showContextMenu();
        }
    },
    ACT_AS() {
        @Override
        void executeOnUiThread(ActorContextMenu menu, NoteEditorData editorData) {
            AccountSelector.selectAccountForActor(menu.getActivity(), menu.menuGroup,
                    ActivityRequestCode.SELECT_ACCOUNT_TO_ACT_AS, menu.getViewItem().getActor());
        }
    },
    FOLLOWERS(true) {
        @Override
        NoteEditorData executeAsync(Params params) {
            setActingAccountForActor(params);
            return super.executeAsync(params);
        }

        @Override
        void executeOnUiThread(ActorContextMenu menu, NoteEditorData editorData) {
            startActorListActivity(menu, ActorListType.FOLLOWERS);
        }
    },
    FRIENDS(true) {
        @Override
        NoteEditorData executeAsync(Params params) {
            setActingAccountForActor(params);
            return super.executeAsync(params);
        }

        @Override
        void executeOnUiThread(ActorContextMenu menu, NoteEditorData editorData) {
            startActorListActivity(menu, ActorListType.FRIENDS);
        }
    },
    NONEXISTENT(),
    UNKNOWN();

    private static class Params {
        ActorContextMenu menu;

        Params(ActorContextMenu menu) {
            this.menu = menu;
        }
    }

    private final boolean mIsAsync;
    private static final String TAG = ActorContextMenuItem.class.getSimpleName();

    ActorContextMenuItem() {
        this(false);
    }

    ActorContextMenuItem(boolean isAsync) {
        this.mIsAsync = isAsync;
    }

    @Override
    public int getId() {
        return Menu.FIRST + ordinal() + 1;
    }
    
    public static ActorContextMenuItem fromId(int id) {
        for (ActorContextMenuItem item : ActorContextMenuItem.values()) {
            if (item.getId() == id) {
                return item;
            }
        }
        return UNKNOWN;
    }

    public void addTo(Menu menu, int menuGroup, int order, int titleRes) {
        menu.add(menuGroup, this.getId(), order, titleRes);
    }

    public void addTo(Menu menu, int menuGroup, int order, CharSequence title) {
        menu.add(menuGroup, this.getId(), order, title);
    }
    
    public boolean execute(ActorContextMenu menu, MyAccount ma) {
        Params params = new Params(menu);
        MyLog.v(this, "execute started");
        if (mIsAsync) {
            executeAsync1(params);
        } else {
            executeOnUiThread(params.menu,
                    new NoteEditorData(menu.getMyContext(),
                            menu.getActingAccount(), 0, 0, false));
        }
        return false;
    }
    
    private void executeAsync1(final Params params) {
        AsyncTaskLauncher.execute(TAG, true,
                new MyAsyncTask<Void, Void, NoteEditorData>(TAG + name(), MyAsyncTask.PoolEnum.QUICK_UI) {
                    @Override
                    protected NoteEditorData doInBackground2(Void... params2) {
                        MyLog.v(this, "execute async started. "
                                + params.menu.getViewItem().actor.getNamePreferablyWebFingerId());
                        return executeAsync(params);
                    }

                    @Override
                    protected void onPostExecute2(NoteEditorData editorData) {
                        MyLog.v(this, "execute async ended");
                        executeOnUiThread(params.menu, editorData);
                    }

                    @Override
                    public String toString() {
                        return TAG + "; " + super.toString();
                    }
                }
        );
    }

    NoteEditorData executeAsync(Params params) {
        return NoteEditorData.newEmpty(params.menu.getActingAccount());
    }

    void executeOnUiThread(ActorContextMenu menu, NoteEditorData editorData) {
        // Empty
    }

    void setActingAccountForActor(Params params) {
        Actor actor = params.menu.getViewItem().getActor();
        Origin origin = params.menu.getOrigin();
        if (!origin.isValid()) {
            MyLog.e(this, "Unknown origin for " + params.menu.getViewItem().actor);
            return;
        }
        if (params.menu.getActingAccount().nonValid() || !params.menu.getActingAccount().getOrigin().equals(origin)) {
            params.menu.setSelectedActingAccount(params.menu.getMyContext().accounts()
                    .fromActorOfSameOrigin(actor));
            if (params.menu.getActingAccount().nonValid()) {
                params.menu.setSelectedActingAccount(params.menu.getMyContext().accounts().
                        getFirstSucceededForOrigin(origin));
            }
        }
    }

    void startActorListActivity(ActorContextMenu menu, ActorListType actorListType) {
        Uri uri = MatchedUri.getActorListUri(
                actorListType,
                menu.getOrigin().getId(),
                menu.getViewItem().getActorId(), "");
        if (MyLog.isVerboseEnabled()) {
            MyLog.d(this, "startFollowersList, uri:" + uri);
        }
        menu.getActivity().startActivity(MyAction.VIEW_FOLLOWERS.getIntent(uri));
    }

    void sendActOnActorCommand(CommandEnum command, ActorContextMenu menu) {
        MyServiceManager.sendManualForegroundCommand(
                CommandData.actOnActorCommand(command, menu.getActingAccount(),
                        menu.getViewItem().getActorId(), menu.getViewItem().actor.getUsername()));
    }
}

File: app/src/main/java/org/andstatus/app/actor/ActorContextMenuItem.kt
/*
 * Copyright (C) 2016 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.andstatus.app.actor

import android.net.Uri
import android.view.Menu
import org.andstatus.app.ActivityRequestCode
import org.andstatus.app.MyAction
import org.andstatus.app.account.AccountSelector
import org.andstatus.app.data.MatchedUri
import org.andstatus.app.list.ContextMenuItem
import org.andstatus.app.note.NoteEditorData
import org.andstatus.app.os.AsyncTaskLauncher
import org.andstatus.app.os.MyAsyncTask
import org.andstatus.app.service.CommandData
import org.andstatus.app.service.CommandEnum
import org.andstatus.app.service.MyServiceManager
import org.andstatus.app.timeline.TimelineActivity.Companion.startForTimeline
import org.andstatus.app.timeline.meta.TimelineType
import org.andstatus.app.util.MyLog

enum class ActorContextMenuItem constructor(private val mIsAsync: Boolean = false) : ContextMenuItem {
    GET_ACTOR(true) {
        override fun executeAsync(params: Params): NoteEditorData {
            params.menu.getViewItem().actor.requestDownload(true)
            return super.executeAsync(params)
        }
    },
    POST_TO(true) {
        override fun executeAsync(params: Params): NoteEditorData {
            return NoteEditorData.newEmpty(params.menu.getActingAccount())
                    .setPublicAndFollowers(false, false)
                    .addActorsBeforeText(mutableListOf(params.menu.getViewItem().actor))
        }

        override fun executeOnUiThread(menu: ActorContextMenu, editorData: NoteEditorData) {
            menu.menuContainer.getNoteEditor()?.startEditingNote(editorData)
        }
    },
    SHARE {
        override fun executeOnUiThread(menu: ActorContextMenu, editorData: NoteEditorData) {
            // TODO
        }
    },
    NOTES_BY_ACTOR(true) {
        override fun executeAsync(params: Params): NoteEditorData {
            startForTimeline(
                    params.menu.getActivity().myContext,
                    params.menu.getActivity(),
                    params.menu.getActivity().myContext.timelines()
                            .forUserAtHomeOrigin(TimelineType.SENT, params.menu.getViewItem().actor)
            )
            return super.executeAsync(params)
        }
    },
    GROUP_NOTES(true) {
        override fun executeAsync(params: Params): NoteEditorData {
            startForTimeline(
                    params.menu.getActivity().myContext,
                    params.menu.getActivity(),
                    params.menu.getActivity().myContext.timelines()
                            .forUserAtHomeOrigin(TimelineType.GROUP, params.menu.getViewItem().actor)
            )
            return super.executeAsync(params)
        }
    },
    FOLLOW {
        override fun executeOnUiThread(menu: ActorContextMenu, editorData: NoteEditorData) {
            sendActOnActorCommand(CommandEnum.FOLLOW, menu)
        }
    },
    STOP_FOLLOWING {
        override fun executeOnUiThread(menu: ActorContextMenu, editorData: NoteEditorData) {
            sendActOnActorCommand(CommandEnum.UNDO_FOLLOW, menu)
        }
    },
    ACT_AS_FIRST_OTHER_ACCOUNT {
        override fun executeOnUiThread(menu: ActorContextMenu, editorData: NoteEditorData) {
            menu.setSelectedActingAccount(menu.getMyContext().accounts()
                    .firstOtherSucceededForSameUser(menu.getViewItem().actor, menu.getActingAccount()))
            menu.showContextMenu()
        }
    },
    ACT_AS {
        override fun executeOnUiThread(menu: ActorContextMenu, editorData: NoteEditorData) {
            AccountSelector.selectAccountForActor(menu.getActivity(), menu.menuGroup,
                    ActivityRequestCode.SELECT_ACCOUNT_TO_ACT_AS, menu.getViewItem().actor)
        }
    },
    FOLLOWERS(true) {
        override fun executeAsync(params: Params): NoteEditorData {
            setActingAccountForActor(params)
            return super.executeAsync(params)
        }

        override fun executeOnUiThread(menu: ActorContextMenu, editorData: NoteEditorData) {
            startActorsScreen(menu, ActorsScreenType.FOLLOWERS)
        }
    },
    FRIENDS(true) {
        override fun executeAsync(params: Params): NoteEditorData {
            setActingAccountForActor(params)
            return super.executeAsync(params)
        }

        override fun executeOnUiThread(menu: ActorContextMenu, editorData: NoteEditorData) {
            startActorsScreen(menu, ActorsScreenType.FRIENDS)
        }
    },
    NONEXISTENT, UNKNOWN;

    class Params(var menu: ActorContextMenu)

    override fun getId(): Int {
        return Menu.FIRST + ordinal + 1
    }

    fun addTo(menu: Menu, menuGroup: Int, order: Int, titleRes: Int) {
        menu.add(menuGroup, this.getId(), order, titleRes)
    }

    fun addTo(menu: Menu, menuGroup: Int, order: Int, title: CharSequence) {
        menu.add(menuGroup, this.getId(), order, title)
    }

    fun execute(menu: ActorContextMenu): Boolean {
        val params = Params(menu)
        MyLog.v(this, "execute started")
        if (mIsAsync) {
            executeAsync1(params)
        } else {
            val myAccount = menu.getActingAccount().getValidOrCurrent(menu.getMyContext())
            val editorData = NoteEditorData(myAccount, 0, false, 0, false)
            executeOnUiThread(params.menu, editorData)
        }
        return false
    }

    private fun executeAsync1(params: Params) {
        AsyncTaskLauncher.execute(TAG,
                object : MyAsyncTask<Void?, Void?, NoteEditorData>(TAG + name, PoolEnum.QUICK_UI) {
                    override fun doInBackground2(aVoid: Void?): NoteEditorData {
                        MyLog.v(this, "execute async started. "
                                + params.menu.getViewItem().actor.getUniqueNameWithOrigin())
                        return executeAsync(params)
                    }

                    override fun onPostExecute2(editorData: NoteEditorData) {
                        MyLog.v(this, "execute async ended")
                        executeOnUiThread(params.menu, editorData)
                    }

                    override fun toString(): String {
                        return TAG + "; " + super.toString()
                    }
                }
        )
    }

    open fun executeAsync(params: Params): NoteEditorData {
        return NoteEditorData.newEmpty(params.menu.getActingAccount())
    }

    open fun executeOnUiThread(menu: ActorContextMenu, editorData: NoteEditorData) {
        // Empty
    }

    fun setActingAccountForActor(params: Params) {
        val actor = params.menu.getViewItem().actor
        val origin = params.menu.getOrigin()
        if (!origin.isValid()) {
            MyLog.w(this, "Unknown origin for " + params.menu.getViewItem().actor)
            return
        }
        if (params.menu.getActingAccount().nonValid || params.menu.getActingAccount().origin != origin) {
            params.menu.setSelectedActingAccount(params.menu.getMyContext().accounts()
                    .fromActorOfSameOrigin(actor))
            if (params.menu.getActingAccount().nonValid) {
                params.menu.setSelectedActingAccount(params.menu.getMyContext().accounts().getFirstPreferablySucceededForOrigin(origin))
            }
        }
    }

    fun startActorsScreen(menu: ActorContextMenu, actorsScreenType: ActorsScreenType) {
        val uri: Uri = MatchedUri.getActorsScreenUri(
                actorsScreenType,
                menu.getOrigin().id,
                menu.getViewItem().getActorId(), "")
        if (MyLog.isDebugEnabled()) {
            MyLog.d(this, "startActorsScreen, $actorsScreenType, uri:$uri")
        }
        menu.getActivity().startActivity(MyAction.VIEW_FOLLOWERS.getIntent(uri))
    }

    fun sendActOnActorCommand(command: CommandEnum, menu: ActorContextMenu) {
        MyServiceManager.sendManualForegroundCommand(
                CommandData.actOnActorCommand(command, menu.getActingAccount(),
                        menu.getViewItem().actor, menu.getViewItem().actor.getUsername()))
    }

    companion object {
        private val TAG: String = ActorContextMenuItem::class.java.simpleName
        fun fromId(id: Int): ActorContextMenuItem {
            for (item in values()) {
                if (item.getId() == id) {
                    return item
                }
            }
            return UNKNOWN
        }
    }
}